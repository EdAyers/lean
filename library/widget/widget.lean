import .html

universe u

meta def widget.mk_tactic_widget {α β σ : Type}
  (init : σ)
  (update : σ → β → tactic (σ × α))
  (view : σ → tactic (html β))
  (error : option format → σ → σ) : component tactic_state α :=
component.mk β (σ × tactic_state)
  (λ ts_new prev, match prev with | (some ⟨s,ts_old⟩) := ⟨s, ts_new⟩ | (none) := ⟨init, ts_new⟩ end)
  (λ _ ⟨s,ts⟩ b,
      match update s b ts with
      | (interaction_monad.result.success ⟨s,a⟩ ts) := ⟨⟨s,ts⟩, some a⟩
      | (interaction_monad.result.exception msg pos _) := ⟨⟨s,ts⟩, none⟩
      end
  )
  (λ _ ⟨s,ts⟩,
      match view s ts with
      | (interaction_monad.result.success x ts) := x
      | (interaction_monad.result.exception msg pos ts) := [html.of_string "view errored"]
      end
  )