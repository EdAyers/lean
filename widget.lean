universe u

inductive attr (α : Type) : Type
| val (name : string) (value : string) : attr
| on_click : (unit → α) → attr
-- | on_mouse : (mouse_event_args → α) → attr
| style : list (string × string) → attr

inductive html (α : Type) : Type
| element : string → list (attr α) → html → html
| of_string : string → html
| empty : html
| append : html→ html→ html

-- | stateful {α β σ : Type} : σ → (β → state σ α) → (σ → widget β) → widget α
-- | memoise
-- |

-- inductive json
-- | mk_array  : list json → json
-- | mk_object : list (string × json) → json
-- | mk_string : string → json
-- | mk_number : int → json
-- | mk_bool : bool → json

structure widget (α : Type) :=
(β : Type)
(σ : Type)
(init   : σ)
(update (state : σ) (action : β) : (σ × α))
(view   (state : σ) : html β)

/- keep everything as concrete as possible but stay open to wacky abstractions.
The component system needs:
- events with actions;
- statefulness
- timers / animation frames? [todo]
- how to interface with tactics? for now just treat it as global state and make `widget (tactic α)`.

How does it work?
When save_widget is called, the vm_obj for that widget is saved in the info manager.

I can index event handlers in 2 ways;
in way 1 I will use the vm index, but I do then need to keep the handler objects alive in the cache by keeping the reference count up.

Issues with thread safety are going to be a problem.-/

-- [todo] replace `widget unit` with `widget update` where update is some editor instruction to fill a hole.
meta constant tactic.save_widget : pos → (tactic_state → widget unit) → tactic unit

/- EXAMPLE: -/

meta def my_tactic := state_t unit tactic
namespace my_tactic
section
local attribute [reducible] my_tactic
meta instance : monad my_tactic := by apply_instance
meta instance : monad_state unit my_tactic := by apply_instance
meta instance : has_monad_lift tactic my_tactic := by apply_instance
end
meta instance (α : Type) : has_coe (tactic α) (my_tactic α) := ⟨monad_lift⟩
open interaction_monad.result
meta def step {α : Type} (m : my_tactic α) : my_tactic unit :=
m >> pure ()
/- istep is used to make an interactive thingy.  -/
meta def istep {α : Type} (line0 col0 line col : nat) (t : my_tactic α) : my_tactic unit :=
⟨λ v s,
  match (@scope_trace _ line col (λ _, t.run v s)) with
  | (success ⟨a,v⟩ s') := success ((),v) s'
  | (exception (some msg) p s') := exception (some msg) (some ⟨line, col⟩) s'
  | (exception none p s') := silent_fail s'
  end
⟩
meta instance : interactive.executor my_tactic :=
{ config_type := unit,
  execute_with := λ n tac, tac.run n >> pure ()
}
meta def hello_world_widget (ts : tactic_state) : widget unit :=
widget.mk unit unit () (λ s b, (s,s)) (λ s, html.of_string "hello world")

meta def save_info (p : pos) : my_tactic unit := do
  tactic.save_info_thunk p (λ _, to_fmt "HELLO HERE IS SOME INFO"),
  -- tactic.save_widget p hello_world_widget,
  pure ()

end my_tactic

example {P Q : Prop} : P → Q → P :=
begin [my_tactic]
  tactic.intros,
  tactic.assumption
end

example {P Q : Prop} : P → Q → P :=
begin
  tactic.intros,
  tactic.assumption
end

#eval "adsf"