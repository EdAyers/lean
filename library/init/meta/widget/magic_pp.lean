prelude
import init.meta.widget.basic
import init.meta.tactic
import init.meta.expr_address

namespace widget

/-- An alternative to format that keeps structural information about the expression.
For lack of a better name I called it `magic`, rename suggestions welcome.

On tag_expr, note that the given `expr` is _not_ necessarily the subexpression of the root expression that `tactic_state.pp_magic` was
called with. For example if the subexpression is under a binder then all of the `expr.var 0`s will be replaced with
a local not in the local context with the name and type set to that of the binder.
-/
meta inductive magic
| tag_expr : expr.address → expr → magic → magic
| compose : magic → magic → magic
| group : magic → magic
| nest : nat → magic → magic
| highlight : format.color → magic → magic
| format : format → magic

open format
meta def magic.to_fmt : magic → format
| (magic.compose m1 m2) := format.compose (magic.to_fmt m1) (magic.to_fmt m2)
| (magic.group m1) := format.group (magic.to_fmt m1)
| (magic.nest i m) := format.nest i $ magic.to_fmt m
| (magic.highlight c m) := format.highlight (magic.to_fmt m) c
| (magic.format f) := f
| (magic.tag_expr ea e m) := magic.to_fmt m

meta instance magic.has_to_fmt : has_to_format magic := ⟨magic.to_fmt⟩

end widget

open widget

/-- A special version of pp which also preserves expression boundary information. -/
meta constant tactic_state.pp_magic   : tactic_state → expr → magic

meta def tactic.pp_magic : expr → tactic magic
| e := tactic.read >>= λ ts, pure $ tactic_state.pp_magic ts e


