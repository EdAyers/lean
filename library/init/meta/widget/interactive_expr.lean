/- Author: E.W.Ayers -/
/- This contains an experimental attempt to get pretty printing to keep expression information so that the user can hover over subterms of an expression in a widget and get information about that subterm.
For example  -/
prelude
import init.meta.widget.magic_pp
import init.meta.widget.tactic_component
import init.meta.tactic
import init.meta.expr_address

meta def subexpr := (expr × expr.address)

namespace widget

open magic
open html attr

def format.color.to_string : format.color → string
| format.color.red := "red" | format.color.green := "green" | format.color.orange := "orange" | format.color.blue := "blue" | format.color.pink := "pink" | format.color.cyan := "cyan" | format.color.grey := "grey"

/-- magic but without any of the formatting stuff like highlighting, groups etc. -/
meta inductive sm : Type
| tag_expr : expr.address → expr → sm → sm
| compose : sm →  sm →  sm
| of_string : string →  sm

meta def to_simple : magic → sm
| (magic.tag_expr ea e m) := sm.tag_expr ea e $ to_simple m
| (magic.group m) := to_simple m
| (magic.nest i m) := to_simple m
| (magic.highlight i m) := to_simple m
| (magic.format f) := sm.of_string $ format.to_string f
| (magic.compose x y) := sm.compose (to_simple x) (to_simple y)

meta def sm.flatten : sm → sm
| (sm.tag_expr e ea m) := (sm.tag_expr e ea $ sm.flatten m)
| (sm.compose x y) :=
  match (sm.flatten x), (sm.flatten y) with
  | (sm.of_string sx), (sm.of_string sy) := sm.of_string (sx ++ sy)
  | (sm.of_string sx), (sm.compose (sm.of_string sy) z) := sm.compose (sm.of_string (sx ++ sy)) z
  | (sm.compose x (sm.of_string sy)), (sm.of_string sz) := sm.compose x (sm.of_string (sy ++ sz))
  | (sm.compose x (sm.of_string sy1)), (sm.compose (sm.of_string sy2) z) := sm.compose x (sm.compose (sm.of_string (sy1 ++ sy2)) z)
  | x, y := sm.compose x y
  end
| (sm.of_string s) := sm.of_string s

namespace interactive_expression

meta inductive action (γ : Type)
| on_mouse_enter : subexpr → action
| on_mouse_leave_all : action
| on_click : subexpr → action
| on_tooltip_action : γ → action
| on_close_tooltip : action

meta def view {γ} (tooltip_component : tc subexpr (action γ)) (click_address : option expr.address) (select_address : option expr.address) :
  subexpr → sm → tactic (list (html (action γ)))
| ⟨ce, current_address⟩ (sm.tag_expr ea e m) := do
  let new_address := current_address ++ ea,
  let select_attrs : list (attr (action γ)) := if some new_address = select_address then [className "highlight"] else [],
  click_attrs  : list (attr (action γ)) ←
    if some new_address = click_address then do
      content ← tc.to_html tooltip_component (e, new_address),
      pure [tooltip $ content]
    else pure [],
  let as := [className "expr-boundary", key (ea)] ++ select_attrs ++ click_attrs,
  inner ← view (e,new_address) m,
  pure [h "span" as inner]
| ca (sm.compose x y) := pure (++) <*> view ca x <*> view ca y
| ca (sm.of_string s) := pure
  [h "span" [
    on_mouse_enter (λ _, action.on_mouse_enter ca),
    on_click (λ _, action.on_click ca),
    key s
  ] [html.of_string s]]


/-- Make an interactive expression. -/
meta def mk {γ} (tooltip : tc subexpr γ) : tc expr γ :=
let tooltip_comp :=
   component.with_props_eq (λ (x y : tactic_state × expr × expr.address), x.2.2 = y.2.2)
   $ component.map_action (action.on_tooltip_action) tooltip in
tc.mk_simple
  (action γ)
  (option subexpr × option subexpr)
  (λ e, pure $ (none, none))
  (λ e ⟨ca, sa⟩ act, pure $
    match act with
    | (action.on_mouse_enter ⟨e, ea⟩) := ((ca, some (e, ea)), none)
    | (action.on_mouse_leave_all)     := ((ca, none), none)
    | (action.on_click ⟨e, ea⟩)       := if some (e,ea) = ca then ((none, sa), none) else ((some (e, ea), sa), none)
    | (action.on_tooltip_action g)    := ((none, sa), some g)
    | (action.on_close_tooltip)       := ((none, sa), none)
    end
  )
  (λ e ⟨ca, sa⟩, do
    ts ← tactic.read,
    let m : sm  := sm.flatten $ to_simple $ tactic_state.pp_magic ts e,
    let m : sm  := sm.tag_expr [] e m, -- [hack] in pp.cpp I forgot to add an expr-boundary for the root expression.
    v ← view tooltip_comp (prod.snd <$> ca) (prod.snd <$> sa) ⟨e, []⟩ m,
    pure $
    [ h "span" [
          className "expr",
          key e.hash,
          on_mouse_leave (λ _, action.on_mouse_leave_all) ] $ v
      -- expr.address.to_string $ match ca with | none := [] | (some ⟨x,y⟩) := y end,
      -- expr.address.to_string $ match sa with | none := [] | (some ⟨x,y⟩) := y end
      ]
  )

/-- Render the implicit arguments for an expression in fancy, little pills. -/
meta def implicit_arg_list (tooltip : tc subexpr empty) (e : expr) : tactic $ html empty := do
  fn ← (mk tooltip) $ expr.get_app_fn e,
  args ← list.mmap (mk tooltip) $ expr.get_app_args e,
  pure $ h "div" []
    ( (h "span" [className "bg-blue br3 ma1 ph2 white"] [fn]) ::
      list.map (λ a, h "span" [className "bg-gray br3 ma1 ph2 white"] [a]) args
    )

meta def type_tooltip : tc subexpr empty :=
tc.stateless (λ ⟨e,ea⟩, do
    y ← tactic.infer_type e,
    y_comp ← mk type_tooltip y,
    implicit_args ← implicit_arg_list type_tooltip e,
    pure [
        h "div" [] [
          h "div" [] [y_comp],
          h "hr" [] [],
          implicit_args,
          h "hr" [] [],
          h "div" [className "gray"] [
            ea.to_string
          ]
        ]
      ]
  )

end interactive_expression

meta def html.of_name {α : Type} : name → html α
| n := html.of_string $ name.to_string n

open tactic

meta def show_type_component : tc expr empty :=
tc.stateless (λ x, do
  y ← infer_type x,
  y_comp ← interactive_expression.mk interactive_expression.type_tooltip $ y,
  pure y_comp
)

meta def tactic_view_component {γ} (local_c : tc expr γ) (target_c : tc expr γ) : tactic (html γ) :=
do
  gs ← get_goals,
  hs : list (html γ) ← gs.mmap (λ g, do
    ts ← read,
    gn ← pp g,
    set_goals [g],
    lcs ← local_context,
    lchs : list (html γ) ← lcs.mmap (λ lc, do
      pn ← pure $ expr.local_pp_name lc,
      lh : html γ ← local_c lc,
      pure
        $ h "tr" [key $ to_string $ expr.local_uniq_name lc] [
            h "td" [] [html.of_name pn],
            h "td" [] [html.of_string " : "],
            h "td" [] [lh]
        ]
    ),
    t_comp ← target_c g,
    (expr.mvar u_n pp_n y) ← pure g,
    pure $ h "table" [key $ expr.hash g, className "font-code"] [
      h "tbody" [] $ lchs ++ [
          h "tr" [key u_n, className "bt"] [
            h "td" [] [] ,
            h "td" [] [html.of_string " ⊢ "],
            h "td" [] [t_comp]
       ]]
    ]
  ),
  set_goals gs,
  pure $ h "ul" [className "list pl0"]
       $ list.mapi (λ i x,
         let border_cn := if i + 1 = hs.length then "ba bl-0 bt-0 br-0 b--dotted b--black-30" else "" in
         h "li" [className $ "lh-copy " ++ border_cn] [x])
       $ hs


meta def tactic_render : tactic (html empty) :=
tactic_view_component show_type_component show_type_component

meta def mk_tactic_widget {α β σ : Type}
  (init : σ)
  (update : σ → β → tactic (σ × α))
  (view : σ → tactic (html β))
  (error : option format → σ → σ) : component tactic_state α :=
component.mk β (σ × tactic_state)
  (λ ts_new prev, match prev with | (some ⟨s,ts_old⟩) := ⟨s, ts_new⟩ | (none) := ⟨init, ts_new⟩ end)
  (λ _ ⟨s,ts⟩ b,
      match update s b ts with
      | (interaction_monad.result.success ⟨s,a⟩ ts) := ⟨⟨s,ts⟩, some a⟩
      | (interaction_monad.result.exception msg pos _) := ⟨⟨s,ts⟩, none⟩
      end
  )
  (λ _ ⟨s,ts⟩,
      match view s ts with
      | (interaction_monad.result.success x ts) := x
      | (interaction_monad.result.exception msg pos ts) := [html.of_string "view errored"]
      end
  )
  (λ x y, ff)

meta def tactic_state_widget : component tactic_state string :=
mk_tactic_widget () (λ _ _, failure) (λ _, tactic_render) (λ _ x, x)

end widget



