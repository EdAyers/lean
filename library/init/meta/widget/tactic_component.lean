import init.meta.tactic
import init.meta.widget.basic

meta def tactic.run_simple {α} : tactic_state → tactic α → option α
| ts t := match t ts with
          | (interaction_monad.result.success a ts') := some a
          | (interaction_monad.result.exception _ _ _) := none
          end

namespace widget

/-- A component that implicitly depends on tactic_state. For efficiency we always assume that the tactic_state is unchanged between component renderings. -/
meta def tc (π : Type) (α : Type) := component (tactic_state × π) (α)

namespace tc

variables {π ρ α β : Type}

meta def of_component : component π α → tc π α :=
component.map_props prod.snd

meta def map_action (f : α → β) : tc π α → tc π β :=
component.map_action f

meta def map_props (f : π → ρ) : tc ρ α → tc π α :=
component.map_props (prod.map id f)


open interaction_monad
open interaction_monad.result

/-- Make a tactic component from some init, update, views which are expecting a tactic.
Note that the resulting tactic state of the view is thrown away but tactic_state is preserved locally on updates.

-/
meta def mk_simple
  [decidable_eq π]
  (β σ : Type)
  (init : π → tactic σ)
  (update : π → σ → β → tactic (σ × option α))
  (view : π → σ → tactic (list (html β)))
  : tc π α :=
@component.mk
  (tactic_state × π)
  α
  β
  (interaction_monad.result tactic_state σ)
  (λ ⟨ts,p⟩ last, match last with (some x) := x | _ := init p ts end)
  (λ ⟨ts,p⟩ s b,
    match s with
    | (success s ts) :=
      match update p s b ts with
      | (success ⟨s,a⟩ ts) := prod.mk (success s ts) (a)
      | (exception m p s) := prod.mk (exception m p s) none
      end
    | x := ⟨x,none⟩
    end
  )
  (λ ⟨ts,p⟩ s,
    match s with
    | (success s ts) :=
      match view p s ts with
      | (success h ts) := h
      | (exception msg pos s) := ["rendering tactic failed "]
      end
    | (exception msg pos s) := ["state of tactic component has failed!"]
    end
  )
  (λ ⟨_,p1⟩ ⟨_,p2⟩, p1 = p2)

-- meta def run : tc unit α → component tactic_state α
-- | c := component.mk _ _ (λ )

end tc

end widget