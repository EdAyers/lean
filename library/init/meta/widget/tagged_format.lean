/-
Copyright (c) E.W.Ayers. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.

Author: E.W.Ayers
-/
prelude
import init.meta.widget.basic
import init.meta.tactic
import init.meta.expr_address

namespace widget

/-- An alternative to format that keeps structural information stored as a tag. -/
meta inductive tagged_format (α : Type)
| tag : α → tagged_format → tagged_format
| compose : tagged_format → tagged_format → tagged_format
| group : tagged_format → tagged_format
| nest : nat → tagged_format → tagged_format
| highlight : format.color → tagged_format → tagged_format
| format : format → tagged_format

variables {α : Type}

open format
protected meta def tagged_format.to_fmt: tagged_format α → format
| (tagged_format.compose m1 m2) := format.compose (tagged_format.to_fmt m1) (tagged_format.to_fmt m2)
| (tagged_format.group m1)      := format.group $ tagged_format.to_fmt m1
| (tagged_format.nest i m)      := format.nest i $ tagged_format.to_fmt m
| (tagged_format.highlight c m) := format.highlight (tagged_format.to_fmt m) c
| (tagged_format.format f)      := f
| (tagged_format.tag a m)       := tagged_format.to_fmt m

meta instance tagged_format.has_to_fmt : has_to_format (tagged_format α) := ⟨tagged_format.to_fmt⟩

end widget

open widget

/-- tagged_format with information about subexpressions. -/
meta def eformat := tagged_format (expr.address × expr)

/-- A special version of pp which also preserves expression boundary information.

On a tag ⟨e,a⟩, note that the given expr `e` is _not_ necessarily the subexpression of the root
expression that `tactic_state.pp_tagged` was called with. For example if the subexpression is
under a binder then all of the `expr.var 0`s will be replaced with a local not in the local context
with the name and type set to that of the binder.-/
meta constant tactic_state.pp_tagged : tactic_state → expr → eformat

meta def tactic.pp_tagged : expr → tactic eformat
| e := tactic.read >>= λ ts, pure $ tactic_state.pp_tagged ts e
