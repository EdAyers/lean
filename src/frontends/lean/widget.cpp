/*
Copyright (c) E.W.Ayers. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.

Author: E.W.Ayers
*/
#include <map>
#include <vector>
#include <string>
#include "library/vm/vm.h"
#include "library/vm/vm_option.h"
#include "library/vm/vm_string.h"
#include "library/vm/vm_list.h"
#include "util/list.h"
#include "frontends/lean/widget.h"
#include "frontends/lean/json.h"
#include "util/optional.h"
#include "util/pair.h"

namespace lean {

optional<std::string> vdom_element::key() {
    if (m_attrs.find("key") != m_attrs.end()) {
        // there is an entry with key "key"
        std::string k = m_attrs["key"];
        return optional<std::string>(k);
    }
    return optional<std::string>();
}
void vdom_element::reconcile(vdom const & old_vdom) {
    vdom_element * o = dynamic_cast<vdom_element*>(old_vdom.raw());
    if (o && o->m_tag == m_tag) {
        std::vector<vdom> ocs = o->m_children;
        reconcile_children(m_children, ocs);
        if (m_tooltip && o->m_tooltip) {
            (*m_tooltip).reconcile(*(o->m_tooltip));
        }
    }
}
json vdom_element::to_json(list<unsigned> const & route) {
    json entry;
    entry["tag"] = m_tag;
    entry["attributes"] = m_attrs;
    json jr = json::array();
    for (auto const & i : route) {
        jr.push_back(i);
    }
    for (auto const & x : m_events) {
        entry["events"][x.first]["route"] = jr;
        entry["events"][x.first]["handler"] = json(x.second);
    }
    entry["children"] = json::array();
    for (vdom & v : m_children) {
        entry["children"].push_back(v.to_json(route));
    }
    if (m_tooltip) {
        entry["tooltip"] = (*m_tooltip).to_json(route);
    }
    return entry;
}

vm_obj component_instance::init(vm_obj const & p, optional<vm_obj> const & s) {
    vm_obj os = s ? mk_vm_some(*s) : mk_vm_none();
    return invoke(cfield(m_component.to_vm_obj(), 0), p, os);
}

pair<vm_obj, optional<vm_obj>> component_instance::update(vm_obj const & p, vm_obj const & s, vm_obj const & a) {
    vm_obj sa = invoke(cfield(m_component.to_vm_obj(), 1), p, s, a);
    vm_obj oa = cfield(sa, 1);
    optional<vm_obj> o = is_none(oa) ? optional<vm_obj>() : optional<vm_obj>(get_some_value(oa));
    return mk_pair(cfield(sa, 0), o);
}

vm_obj component_instance::view(vm_obj const & p, vm_obj const & s) {
    return invoke(cfield(m_component.to_vm_obj(), 2), p, s);
}

void component_instance::render() {
    std::vector<component_instance *> children;
    std::map<unsigned, ts_vm_obj> handlers;
    std::vector<vdom> elements = render_html_list(view(m_props.to_vm_obj(), (*m_state).to_vm_obj()), children, handlers, cons(m_id, m_route));
    std::vector<vdom> old_elements = m_render;
    reconcile_children(elements, old_elements);
    m_handlers = handlers;
    m_children = children;
    m_render = elements;
    m_has_rendered = true;
}

void component_instance::reconcile(vdom const & old) {
    lean_assert(!m_has_rendered);
    component_instance * ci_old = dynamic_cast<component_instance *>(old.raw());
    if (ci_old->m_component.to_vm_obj().raw() == m_component.to_vm_obj().raw()) { // [todo] check this works
        // copy over the state from our previous version.
        if (m_props.to_vm_obj() == ci_old->m_props.to_vm_obj()) { // [todo] use dec-eq here
            m_handlers = ci_old->m_handlers;
            m_children = ci_old->m_children;
            m_render   = ci_old->m_render;
            m_state    = ci_old->m_state;
            m_id       = ci_old->m_id;
            m_has_rendered = true;
            lean_assert(m_route == ci_old->m_route);
        } else {
            optional<vm_obj> old_state = some((*(ci_old->m_state)).to_vm_obj());
            ts_vm_obj new_state = init(m_props.to_vm_obj(), old_state);
            m_state = optional<ts_vm_obj>(new_state);
            render();
        }
    } else {
        m_state = some<ts_vm_obj>(init(m_props.to_vm_obj(), optional<vm_obj>()));
        render();
    }
}

json component_instance::to_json(list<unsigned> const & route) {
    if (!m_has_rendered) {
        if (!m_state) {
            m_state = some<ts_vm_obj>(init(m_props.to_vm_obj(), optional<vm_obj>()));
        }
        render();
    }
    json result = json::array();
    for (vdom const & x : m_render) {
        result.push_back(x.to_json(cons(m_id, route)));
    }
    return result;
}

optional<vm_obj> component_instance::handle_action(vm_obj const & a) {
    auto p = update(m_props.to_vm_obj(), (*m_state).to_vm_obj(), a);
    m_state = p.first;
    render();
    return p.second;
}

optional<vm_obj> component_instance::handle_event(list<unsigned> const & route, unsigned handler_id, vm_obj const & event_args) {
    if (empty(route)) {
        if (m_handlers.find(handler_id) == m_handlers.end()) {
            // component may have rerendered, but handler_id refers to event handler on older component.
            throw invalid_handler();
        }
        ts_vm_obj handler = m_handlers[handler_id];
        // [todo] to prevent a VM error in the case of bad client code, we should check that the 'type' of the event_args here is the same as what the handler expects.
        // the event record from the client should have a `type` member which can be checked. So each of `m_handlers` should also include a string 'type' to check against.
        auto action = invoke(handler.to_vm_obj(), event_args);
        return handle_action(action);
    }
    for (auto const & c : m_children) {
        if (c->m_id == head(route)) {
            optional<vm_obj> a = c->handle_event(tail(route), handler_id, event_args);
            if (a) {
                return handle_action(*a);
            } else {
                return optional<vm_obj>();
            }
        }
    }
    // given component no longer exists. This happens if the ui is updated but there are events from an old
    throw invalid_handler();
}

void reconcile_children(std::vector<vdom> & new_elements, std::vector<vdom> const & olds) {
    std::vector<vdom> old_elements = olds;
    for (unsigned i = 0; i < new_elements.size(); i++) {
        // [note] you get wierd behaviour if multiple things have the same key or if only some elements have keys
        // but this is also true in React so I am not too worried about it as long as it doesn't crash.
        // [note] could probably avoid a few vdom copies but w/e
        auto k = new_elements[i].key();
        if (k) {
            for (unsigned j = 0; j < old_elements.size(); j++) {
                if (old_elements[j].key() == k) {
                    vdom o = old_elements[j];
                    new_elements[i].reconcile(o);
                    old_elements.erase(old_elements.begin() + j);
                    break;
                }
            }
        } else if (old_elements.size() > 0) {
            new_elements[i].reconcile(old_elements[0]);
            old_elements.erase(old_elements.begin());
        } else {
            // continue
        }
    }
}

void render_event(std::string const & name, vm_obj const & handler, std::map<std::string, unsigned> & events, std::map<unsigned, ts_vm_obj> & handlers) {
    static unsigned handler_count = 0; // [fixme] threading issues here;
    unsigned handler_id = handler_count++;

    events[name] = handler_id;
    handlers[handler_id] = handler;
}

vdom render_element(vm_obj const & elt, std::vector<component_instance*> & components, std::map<unsigned, ts_vm_obj> & handlers, list<unsigned> const & route) {
    std::string tag = to_string(cfield(elt, 0));
    vm_obj v_attrs = cfield(elt, 1);
    json attributes;
    std::map<std::string, unsigned> events;
    optional<vdom> tooltip;
    while (!is_simple(v_attrs)) {
        vm_obj attr = head(v_attrs);
        v_attrs = tail(v_attrs);
        switch (cidx(attr)) {
            case 0: { // val : string -> string -> attr
                std::string key = to_string(cfield(attr, 0));
                std::string value = to_string(cfield(attr, 1));
                if (key == "className" && attributes.find(key) != attributes.end()) { // [hack] className fields should be merged. Should really be handled in lean.
                    std::string cn = attributes[key];
                    cn += " ";
                    cn += value;
                    attributes[key] = cn;
                } else {
                    attributes[key] = value;
                }
                break;
            }
            case 1: {// on_mouse_event : mouse_event_kind -> (unit -> Action) -> html.attr
                switch (cidx(cfield(attr, 0))) {
                    case 0: render_event("onClick",      cfield(attr, 1), events, handlers); break;
                    case 1: render_event("onMouseEnter", cfield(attr, 1), events, handlers); break;
                    case 2: render_event("onMouseLeave", cfield(attr, 1), events, handlers); break;
                    default: lean_unreachable();
                }
                break;
            }
            case 2: { // style : list (string × string) → html.attr
                auto l = cfield(attr, 0);
                while (!is_simple(l)) {
                    auto h = head(l);
                    attributes["style"][to_string(cfield(h, 0))] = to_string(cfield(h, 1));
                    l = tail(l);
                }
                break;
            }
            case 3: { // tooltip : html Action → html.attr
                vdom tooltip_child = render_html(cfield(attr, 0), components, handlers, route);
                tooltip = optional<vdom>(tooltip_child);
                break;
            }
            case 4: { // text_change_event : (string -> Action) -> html.attr
                render_event("onChange", cfield(attr, 0), events, handlers);
                break;
            }
            default : {
                lean_unreachable();
                break;
            }
        }
    }
    std::vector<vdom> children = render_html_list(cfield(elt, 2), components, handlers, route);
    return vdom(new vdom_element(tag, attributes, events, children, tooltip));
}

vdom render_html(vm_obj const & html, std::vector<component_instance*> & components, std::map<unsigned, ts_vm_obj> & handlers, list<unsigned> const & route) {
    switch (cidx(html)) {
        case 0: { // of_element : element -> html
            vdom elt = render_element(cfield(html, 0), components, handlers, route);
            return elt;
        }
        case 1: { // of_string : string -> html
            return vdom(new vdom_string(to_string(cfield(html, 0))));
        }
        case 2: { // of_component : {π : Type}: π → component π α → html
            vm_obj props = cfield(html, 0);
            vm_obj comp = cfield(html, 1);
            component_instance * c = new component_instance(comp, props, route);
            components.push_back(c);
            return vdom(c);
        }
        default: {
            lean_unreachable();
        }
    }
}

std::vector<vdom> render_html_list(vm_obj const & htmls, std::vector<component_instance*> & components, std::map<unsigned, ts_vm_obj> & handlers, list<unsigned> const & route) {
    std::vector<vdom> elements;
    vm_obj l = htmls;
    while (!is_simple(l)) {
        vdom x = render_html(head(l), components, handlers, route);
        elements.push_back(x);
        l = tail(l);
    }
    return elements;
}

// class vm_vdom : public vm_external {
//     vdom m_val;
//     vm_vdom(vdom const & v) : m_val(v) {}
//     virtual ~vm_vdom() {}
//     virtual void dealloc() { this->~vm_vdom(); get_vm_allocator().deallocate(sizeof(vm_vdom), this); }
//     virtual vm_external * ts_clone(vm_clone_fn const &) override { return new vm_vdom(m_val); }
//     virtual vm_external * clone(vm_clone_fn const &) override { return new (get_vm_allocator().allocate(sizeof(vm_vdom))) vm_vdom(m_val); }
// }

// vm_obj to_obj(vdom const & v) {
//     return mk_vm_external(new(get_vm_allocator().allocate(sizeof(vm_vdom))) vm_vdom(v));
// }

// vdom to_vdom(vm_obj const & o) {
//     lean_vm_check(dynamic_cast<vm_vdom*>(to_external(o)));
//     return static_cast<vm_vdom*>(to_external(o))->m_val;
// }

vm_obj vm_of_element(vm_obj const &, vm_obj const & e) {
    return mk_vm_constructor(0, e);
}

vm_obj vm_of_string(vm_obj const &, vm_obj const & o) {
    return mk_vm_constructor(1, o);
}

vm_obj vm_of_component(vm_obj const &, vm_obj const &, vm_obj const & vprops, vm_obj const & vcomp) {
    return mk_vm_constructor(2, vprops, vcomp);
}

vm_obj vm_html_cases(vm_obj const &, vm_obj const &, vm_obj const & fe, vm_obj const & fs, vm_obj const & fc, vm_obj const & html) {
    switch (cidx(html)) {
        case 0: return invoke(fe, cfield(html, 0));
        case 1: return invoke(fs, cfield(html, 0));
        case 2: return invoke(fc, mk_vm_simple(0), cfield(html, 0), cfield(html, 1));
        default: lean_unreachable();
    }
}

vm_obj vm_component_mk(
    vm_obj const &, vm_obj const &,
    vm_obj const &, vm_obj const &,
    vm_obj const & i,
    vm_obj const & u,
    vm_obj const & v) {
    return mk_vm_constructor(0, i, u, v);
    }

void initialize_widget() {
    DECLARE_VM_BUILTIN(name({"html", "of_element"}), vm_of_element);
    DECLARE_VM_BUILTIN(name({"html", "of_string"}),  vm_of_string);
    DECLARE_VM_BUILTIN(name({"html", "of_component"}), vm_of_component);
    DECLARE_VM_BUILTIN(name({"html", "cases"}),  vm_html_cases);

    DECLARE_VM_BUILTIN(name({"component", "mk"}), vm_component_mk);
    DECLARE_VM_BUILTIN(name({"component", "state"}),   [](vm_obj const &, vm_obj const &) { return mk_vm_simple(0); });
    DECLARE_VM_BUILTIN(name({"component", "event"}),   [](vm_obj const &, vm_obj const &) { return mk_vm_simple(0); });
    DECLARE_VM_BUILTIN(name({"component", "init"}),    [](vm_obj const &, vm_obj const &, vm_obj const & c) { return cfield(c, 0); });
    DECLARE_VM_BUILTIN(name({"component", "update"}),  [](vm_obj const &, vm_obj const &, vm_obj const & c) { return cfield(c, 1); });
    DECLARE_VM_BUILTIN(name({"component", "view"}),    [](vm_obj const &, vm_obj const &, vm_obj const & c) { return cfield(c, 2); });
}

void finalize_widget() {}

}
