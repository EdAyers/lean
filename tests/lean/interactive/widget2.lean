import widget

example {P Q : Prop} : P → Q → P ∧ Q :=
begin [widget_tactic]
  put (component.counter_widget),
  tactic.intros,
  tactic.split,
  tactic.assumption, tactic.assumption,
end