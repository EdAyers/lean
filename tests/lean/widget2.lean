import widget

#eval html.cases (λ e, 0) (λ e, 1) (λ _ p c, 2) $ @html.of_string unit "hello"

#eval option.map html.element.tag $ @html.as_element unit $ html.div $ [html.of_string "hello"]
