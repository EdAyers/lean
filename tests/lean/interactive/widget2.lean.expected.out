{"message":"file invalidated","response":"ok","seq_num":0}
